### PR TITLE
[receiver/kafkareceiver] handles Kafka rebalancing impact

### LINKED ISSUE #17312: [receiver/kafkareceiver] Consider Kafka rebalancing while consuming messages
### Component(s)

receiver/kafka

### Describe the issue you're reporting

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/951245d4273845de59db3b00f5445bcad9ce4372/receiver/kafkareceiver/kafka_receiver.go#L412

`ConsumeClaim` defined in Kafka Receiver does not take rebalancing into consideration. `ConsumeClaim` should watch the `sarama.ConsumerGroupSession.Context().Done()` and return when context is marked as done.

### PR DESCRIPTION
**Pull request by avadhut123pisal**
**Description:** <Describe what has changed.>

Watches the `sarama.ConsumerGroupSession.Context().Done()` in `consumeClaim` to return when context is marked as done. Closes #17312 .

**Link to tracking Issue:** #17312 

**Testing:** <Describe what testing was performed and which tests were added.>

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

**Comment by avadhut123pisal** [comment]
Hi @pavolloffay / @evan-bradley, will you please go through the changes ?

**Review by evan-bradley — COMMENTED** [review]
Thanks for fixing this. Could you add or update the tests for this receiver to verify the intended behavior?

**Comment by avadhut123pisal** [comment]
> Thanks for fixing this. Could you add or update the tests for this receiver to verify the intended behavior?

Added a test case for the same.

File: .chloggen/handle-kafka-rebalancing-impact-in-consumer.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: enhancement
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: kafkareceiver
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Handle an impact of kafka rebalancing.
```
[the remark this answers is not in the document]
Comment (avadhut123pisal) [reply .chloggen/handle-kafka-rebalancing-impact-in-consumer.yaml]: done

**Review by evan-bradley — COMMENTED** [review]
Thanks for adding the test. Mostly looks good, just a few more requests.

File: receiver/kafkareceiver/kafka_receiver_test.go [hunk 2]
```diff
@@ -177,6 +177,57 @@ func TestTracesConsumerGroupHandler(t *testing.T) {
 	wg.Wait()
 }
 
+func TestTracesConsumerGroupHandler_session_done(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (avadhut123pisal) [reply receiver/kafkareceiver/kafka_receiver_test.go:180]: done.

**Review by evan-bradley — APPROVED** [review]
Looks good to me. Thank you for taking initiative on this issue.

**Review by MovieStoreGuy — APPROVED** [review]
This is awesome! Something I've wanted to tackle for a long time.

**Pull request merged**